{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/pachterlab/HCP_2024/blob/main/cis_trans_Tsouris_GC.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7jGm5WQZjVxf",
        "outputId": "59d12118-725b-4916-e319-86df6505a532"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "--2024-04-26 23:53:50--  http://1002genomes.u-strasbg.fr/files/Diallel_RNAseq/ASE/Datafile2_ase_sum_20230609.tab\n",
            "Resolving 1002genomes.u-strasbg.fr (1002genomes.u-strasbg.fr)... 130.79.18.41\n",
            "Connecting to 1002genomes.u-strasbg.fr (1002genomes.u-strasbg.fr)|130.79.18.41|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 134196476 (128M) [text/plain]\n",
            "Saving to: ‘Datafile2_ase_sum_20230609.tab’\n",
            "\n",
            "Datafile2_ase_sum_2 100%[===================>] 127.98M  26.1MB/s    in 8.1s    \n",
            "\n",
            "2024-04-26 23:53:59 (15.7 MB/s) - ‘Datafile2_ase_sum_20230609.tab’ saved [134196476/134196476]\n",
            "\n"
          ]
        }
      ],
      "source": [
        "!wget http://1002genomes.u-strasbg.fr/files/Diallel_RNAseq/ASE/Datafile2_ase_sum_20230609.tab"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!mkdir results\n",
        "!mkdir figs"
      ],
      "metadata": {
        "id": "Rd_xezwHyq0B"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import numpy as np\n",
        "import pandas as pd\n",
        "from scipy import stats\n",
        "np.random.seed(307203)\n",
        "\n",
        "# plotting\n",
        "import matplotlib\n",
        "import matplotlib.pyplot as plt"
      ],
      "metadata": {
        "id": "Ip3GlGfZXob6"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df = pd.read_csv('Datafile2_ase_sum_20230609.tab')"
      ],
      "metadata": {
        "id": "C4N5_6P7Xon3"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "-----\n",
        "# Hypothesis testing\n",
        "\n",
        "## Binomial test for $H1 = H2$"
      ],
      "metadata": {
        "id": "2kd5KCMWyyDD"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "N = len(df)\n",
        "# N = 10000\n",
        "\n",
        "binom_pvals = np.empty(N)\n",
        "\n",
        "\n",
        "for i in range(N):\n",
        "  print(i)\n",
        "#   if (i^100) == 0:\n",
        "#     print(i)\n",
        "  parent_1 = int(np.round(df.loc[i,'P1_2n_count']))\n",
        "  parent_2 = int(np.round(df.loc[i,'P2_2n_count']))\n",
        "  hybrid_1 = int(np.round(df.loc[i,'P1_count']))\n",
        "  hybrid_2 = int(np.round(df.loc[i,'P2_count']))\n",
        "\n",
        "  result_binom = stats.binomtest(hybrid_1, n=hybrid_1 + hybrid_2, p=0.5, alternative='two-sided')\n",
        "  binom_pvals[i] = result_binom.pvalue\n",
        "\n",
        "\n",
        "\n",
        "np.save('./results/binom_pvals_Tsouris',binom_pvals)\n"
      ],
      "metadata": {
        "id": "sLYnIASsXpfG"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "## Bivariate binomial test for $\\frac{P1}{P1+P2} = \\frac{H1}{H1+H2}$"
      ],
      "metadata": {
        "id": "WecxYnJ2y2gX"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "N = len(df)\n",
        "# N = 100\n",
        "print(N)\n",
        "\n",
        "binom_pvals = np.empty(N)\n",
        "bi_binom_pvals = np.empty(N)\n",
        "\n",
        "\n",
        "for i in range(N):\n",
        "  print(i)\n",
        "#   if (i^100) == 0:\n",
        "#     print(i)\n",
        "  parent_1 = int(np.round(df.loc[i,'P1_2n_count']))\n",
        "  parent_2 = int(np.round(df.loc[i,'P2_2n_count']))\n",
        "  hybrid_1 = int(np.round(df.loc[i,'P1_count']))\n",
        "  hybrid_2 = int(np.round(df.loc[i,'P2_count']))\n",
        "\n",
        "  p_theoretical = (parent_1+hybrid_1)/(parent_1+parent_2+hybrid_1+hybrid_2)\n",
        "\n",
        "  p_obs_p = stats.binom.pmf(k = parent_1,\n",
        "                            n = parent_1+parent_2,\n",
        "                            p = p_theoretical)\n",
        "  p_obs_h = stats.binom.pmf(k = hybrid_1,\n",
        "                            n = hybrid_1+hybrid_2,\n",
        "                            p = p_theoretical)\n",
        "  p_obs = p_obs_p*p_obs_h\n",
        "\n",
        "\n",
        "  prob_P_array = stats.binom.pmf(k = np.arange(0,parent_1+parent_2+1),\n",
        "                                 n = parent_1+parent_2 ,\n",
        "                                 p = p_theoretical).reshape(-1,1).repeat(hybrid_1+hybrid_2+1,axis=1)\n",
        "  prob_H_array = stats.binom.pmf(k = np.arange(0,hybrid_1+hybrid_2+1),\n",
        "                                 n = hybrid_1+hybrid_2,\n",
        "                                 p = p_theoretical).reshape(1,-1).repeat(parent_1+parent_2+1,axis=0)\n",
        "\n",
        "  prob_array = prob_P_array*prob_H_array\n",
        "  p_sum = np.sum(prob_array[prob_array<=p_obs])\n",
        "  bi_binom_pvals[i] = p_sum\n",
        "\n",
        "\n",
        "\n",
        "np.save('./results/binom_ratio_pvals_Tsouris',bi_binom_pvals)\n"
      ],
      "metadata": {
        "id": "Ud33kcvvXplH"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Raw p-value distributions"
      ],
      "metadata": {
        "id": "QC039w_Oy-Px"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "binom_pvals = np.load('./results/binom_pvals_Tsouris.npy')\n",
        "bi_binom_pvals = np.load('./results/binom_ratio_pvals_Tsouris.npy')\n",
        "\n",
        "df['binom_pvals_uncorrected'] = binom_pvals\n",
        "df['bi_binom_pvals_uncorrected'] = bi_binom_pvals"
      ],
      "metadata": {
        "id": "Qi8AQ_lmXpwC"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# what do the p-value distributions look like?\n",
        "fs = 20\n",
        "fig,ax = plt.subplots(1,2,figsize=(10,5))\n",
        "\n",
        "ax[0].hist(binom_pvals,density=False,bins = 45)\n",
        "ax[0].set_title('Binomial test (null = $trans$)',fontsize=fs)\n",
        "ax[0].set_xlabel('Uncorrected p-values',fontsize=fs)\n",
        "ax[0].set_ylabel('Number of genes',fontsize=fs)\n",
        "ax[0].grid()\n",
        "\n",
        "ax[1].hist(bi_binom_pvals,bins = 45,density=False)\n",
        "ax[1].set_title('Binomial ratio test (null = $cis$)',fontsize=fs)\n",
        "ax[1].set_xlabel('Uncorrected p-values',fontsize=fs)\n",
        "ax[1].set_ylabel('Number of genes',fontsize=fs)\n",
        "ax[1].grid()\n",
        "\n",
        "plt.tight_layout()\n",
        "plt.savefig('./figs/uncorrected_pvals',dpi=450,bbox_inches='tight')"
      ],
      "metadata": {
        "id": "God0FP-LzAhe"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Correct for multiple testing using Benjamini-Hochberg."
      ],
      "metadata": {
        "id": "2bEXU7oIzEum"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "df['sample-gene'] = df['sampleID']+df['GeneID']\n",
        "sample_gene = df['sample-gene'].values.flatten()\n",
        "\n",
        "num_test = len(binom_pvals)\n",
        "\n",
        "# sort p values\n",
        "sorted_binom_genes = [g for _,g in sorted(zip(binom_pvals,sample_gene))]\n",
        "sorted_binom_pvals = [p for p,_ in sorted(zip(binom_pvals,sample_gene))]\n",
        "\n",
        "# calculate FDR for each gene\n",
        "fdr_binom = np.arange(1,num_test+1)/num_test * sorted_binom_pvals\n",
        "\n",
        "# create df\n",
        "df_binom = pd.DataFrame({'fdr_binom' : sorted_binom_pvals, 'sample-gene' : sorted_binom_genes})\n",
        "\n",
        "# sort p values\n",
        "sorted_bi_binom_genes = [g for _,g in sorted(zip(bi_binom_pvals,sample_gene))]\n",
        "sorted_bi_binom_pvals = [p for p,_ in sorted(zip(bi_binom_pvals,sample_gene))]\n",
        "\n",
        "# calculate FDR for each gene\n",
        "fdr_bi_binom = np.arange(1,num_test+1)/num_test * sorted_bi_binom_pvals\n",
        "\n",
        "# create df\n",
        "df_bi_binom = pd.DataFrame({'fdr_bi_binom' : sorted_bi_binom_pvals, 'sample-gene' : sorted_bi_binom_genes})\n",
        "\n",
        "# merge\n",
        "df_binom = df_binom.merge(df_bi_binom,on='sample-gene')\n",
        "\n",
        "df = df.merge(df_binom,on='sample-gene')\n"
      ],
      "metadata": {
        "id": "MesVqCZqzDk7"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Corrected p-value (FDR) distributions"
      ],
      "metadata": {
        "id": "6Thrf69jzIG5"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# histogram of corrected p values\n",
        "\n",
        "# what do the p-value distributions look like?\n",
        "fs = 20\n",
        "fig,ax = plt.subplots(1,2,figsize=(10,5))\n",
        "\n",
        "ax[0].hist(fdr_binom,density=False,bins = 45,color = 'darkred')\n",
        "ax[0].set_title('Binomial test (null = $trans$)',fontsize=fs)\n",
        "ax[0].set_xlabel('False discovery rates (FDR)',fontsize=fs)\n",
        "ax[0].set_ylabel('Number of genes',fontsize=fs)\n",
        "ax[0].grid()\n",
        "\n",
        "ax[1].hist(fdr_bi_binom,bins = 45,density=False, color = 'darkred')\n",
        "ax[1].set_title('Binomial ratio test (null = $cis$)',fontsize=fs)\n",
        "ax[1].set_xlabel('False discovery rates (FDR)',fontsize=fs)\n",
        "ax[1].set_ylabel('Number of genes',fontsize=fs)\n",
        "ax[1].grid()\n",
        "\n",
        "plt.tight_layout()\n",
        "plt.savefig('./figs/corrected_fdr',dpi=450,bbox_inches='tight')"
      ],
      "metadata": {
        "id": "-6bN1GP8zJXI"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Gene regulatory assignments based on FDRs"
      ],
      "metadata": {
        "id": "ugF9rG5ezL-p"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "FDR = 0.05\n",
        "\n",
        "# Assign our new regulatory groups\n",
        "P = df['Parlog2FC'].values.flatten()\n",
        "H = df['Hyblog2FC'].values.flatten()\n",
        "delta = P - H\n",
        "\n",
        "ordered_fdr_ztest = df['bi_binom_pvals_uncorrected'].values.flatten()\n",
        "ordered_fdr_binom = df['binom_pvals_uncorrected'].values.flatten()\n",
        "\n",
        "\n",
        "# make OUR new regulation assignments:\n",
        "new_reg_assignments = np.empty(len(delta), dtype=object)\n",
        "\n",
        "# cis\n",
        "index = (ordered_fdr_binom<=FDR) & (ordered_fdr_ztest>FDR)\n",
        "new_reg_assignments[index] = 'cis'\n",
        "\n",
        "# trans\n",
        "index = (ordered_fdr_binom>FDR) & (ordered_fdr_ztest<=FDR)\n",
        "new_reg_assignments[index] = 'trans'\n",
        "\n",
        "# cis x trans\n",
        "index = (ordered_fdr_binom<=FDR) & (ordered_fdr_ztest<=FDR) & (delta*H<=0)\n",
        "new_reg_assignments[index&index] = 'cis x trans'\n",
        "\n",
        "# cis + trans\n",
        "index = (ordered_fdr_binom<=FDR) & (ordered_fdr_ztest<=FDR) & (delta*H>=0)\n",
        "new_reg_assignments[index&index] = 'cis + trans'\n",
        "\n",
        "# conserved\n",
        "index = (ordered_fdr_binom>FDR) & (ordered_fdr_ztest>FDR)\n",
        "new_reg_assignments[index] = 'conserved'\n",
        "\n",
        "\n",
        "df['new_reg_group_fdr'] = new_reg_assignments"
      ],
      "metadata": {
        "id": "gSJecOEnzPn4"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "-----\n",
        "# Comparing reported and transformed regulatory assignments"
      ],
      "metadata": {
        "id": "diJOVWFBzQ3X"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# rename the NaN regulatory group to 'Undefined'\n",
        "df['reg_group'] = df['reg_group'].fillna('conserved')\n",
        "reg_groups = df['reg_group'].unique()"
      ],
      "metadata": {
        "id": "H6TMXNTfzSkZ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "color_dict = {'cis' : 'orangered',\n",
        "              'cis + trans' : 'skyblue',\n",
        "              'cis x trans' : 'forestgreen',\n",
        "              'trans' : 'royalblue',\n",
        "               'conserved' : 'lightgray'}\n",
        "label_dict = {'cis' : '$cis$ only',\n",
        "              'cis + trans' : 'cis + trans',\n",
        "              'cis x trans' : 'cis x trans',\n",
        "              'trans' : '$trans$ only',\n",
        "               'conserved' : 'null'}\n"
      ],
      "metadata": {
        "id": "X9vEEOqHzUjo"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# recode to make four possible reg groups for OLD method\n",
        "\n",
        "P = df['Parlog2FC'].values.flatten()\n",
        "H = df['Hyblog2FC'].values.flatten()\n",
        "\n",
        "delta = P - H\n",
        "scaled_theta = (2/np.pi) * np.arctan( H / (delta) ) # scaled theta\n",
        "R = np.sqrt(delta**2  + H**2)\n",
        "df['delta'] = delta\n",
        "df['theta'] = np.arctan( H / (delta) )\n",
        "df['theta_scaled'] = (2/np.pi) * np.arctan( H / (delta) )\n",
        "df['R'] = np.sqrt( (delta) **2 + H**2)\n",
        "\n",
        "\n",
        "# recode THEIR reg groups\n",
        "new_reg_assignments_theirs = np.empty(len(scaled_theta), dtype=object)\n",
        "\n",
        "# cis + trans\n",
        "index = ((delta > 0) & (H > 0)) | ( (delta < 0) & (H < 0) )\n",
        "new_reg_assignments_theirs[index] = 'cis + trans'\n",
        "\n",
        "# cis x trans\n",
        "index = ((delta <= 0) & (H >= 0)) | ( (delta >= 0) & (H <= 0) )\n",
        "new_reg_assignments_theirs[index] = 'cis x trans'\n",
        "\n",
        "\n",
        "# cis\n",
        "index = df['reg_group'].values == 'cis'\n",
        "new_reg_assignments_theirs[index] = 'cis'\n",
        "\n",
        "# trans\n",
        "index = df['reg_group'].values == 'trans'\n",
        "new_reg_assignments_theirs[index] = 'trans'\n",
        "\n",
        "# null\n",
        "index = df['reg_group'].values == 'conserved'\n",
        "new_reg_assignments_theirs[index] =  'conserved'\n",
        "\n",
        "df['new_reg_group_theirs'] = new_reg_assignments_theirs"
      ],
      "metadata": {
        "id": "84fLTL2SzWVh"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# recode to make four possible reg groups for OLD method\n",
        "\n",
        "P = df['Parlog2FC'].values.flatten()\n",
        "H = df['Hyblog2FC'].values.flatten()\n",
        "\n",
        "delta = P - H\n",
        "scaled_theta = (2/np.pi) * np.arctan( H / (delta) ) # scaled theta\n",
        "R = np.sqrt(delta**2  + H**2)\n",
        "df['delta'] = delta\n",
        "df['theta'] = np.arctan( H / (delta) )\n",
        "df['theta_scaled'] = (2/np.pi) * np.arctan( H / (delta) )\n",
        "df['R'] = np.sqrt( (delta) **2 + H**2)\n",
        "\n",
        "\n",
        "# recode THEIR reg groups\n",
        "new_reg_assignments_theirs = np.empty(len(scaled_theta), dtype=object)\n",
        "\n",
        "# cis + trans\n",
        "index = ((delta > 0) & (H > 0)) | ( (delta < 0) & (H < 0) )\n",
        "new_reg_assignments_theirs[index] = 'cis + trans'\n",
        "\n",
        "# cis x trans\n",
        "index = ((delta <= 0) & (H >= 0)) | ( (delta >= 0) & (H <= 0) )\n",
        "new_reg_assignments_theirs[index] = 'cis x trans'\n",
        "\n",
        "\n",
        "# cis\n",
        "index = df['reg_group'].values == 'cis'\n",
        "new_reg_assignments_theirs[index] = 'cis'\n",
        "\n",
        "# trans\n",
        "index = df['reg_group'].values == 'trans'\n",
        "new_reg_assignments_theirs[index] = 'trans'\n",
        "\n",
        "# null\n",
        "index = df['reg_group'].values == 'conserved'\n",
        "new_reg_assignments_theirs[index] =  'conserved'\n",
        "\n",
        "df['new_reg_group_theirs'] = new_reg_assignments_theirs\n",
        "\n",
        "\n",
        "## Reported regulatory assignments\n",
        "\n",
        "# add column of color to dataframe\n",
        "df['color_reported'] = [color_dict[reg_] for reg_ in df['new_reg_group_theirs'].values]\n",
        "df['color_fdr'] = [color_dict[reg_] for reg_ in df['new_reg_group_fdr'].values]"
      ],
      "metadata": {
        "id": "EscYS5rwz5Yz"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "reg_groups = ['cis x trans','cis + trans','trans','cis','conserved']\n",
        "\n",
        "# random order\n",
        "fig, ax = plt.subplots(3, 1, figsize=(6, 23*(3/4)))\n",
        "\n",
        "s = 2\n",
        "alpha = 1.0\n",
        "axis_fs = 25\n",
        "title_fs = 18\n",
        "label_fs = 20\n",
        "line_color = 'black'\n",
        "axis_width = 3.0\n",
        "line_alpha = 0.7\n",
        "\n",
        "P = df['Parlog2FC'].values.flatten()\n",
        "H = df['Hyblog2FC'].values.flatten()\n",
        "delta = P - H\n",
        "theta_scaled = (2/np.pi) * np.arctan(H / delta)\n",
        "R = np.sqrt(delta**2 + H**2)\n",
        "cis_prop_reordered = theta_scaled-0.5\n",
        "cis_prop_reordered[cis_prop_reordered<=-1.0] = (cis_prop_reordered[cis_prop_reordered<=-1.0]) + 2.0\n",
        "\n",
        "colors = df['color_reported'].values\n",
        "\n",
        "random_idx = np.arange(0,len(colors))\n",
        "np.random.shuffle(random_idx)\n",
        "\n",
        "\n",
        "ax[0].scatter(P[random_idx],H[random_idx],s = s, alpha = alpha, c = colors[random_idx])\n",
        "ax[0].set_xlabel('$R_{P}$ (log2 parental fold change)', fontsize = axis_fs)\n",
        "ax[0].set_ylabel('$R_{H}$ (log2 hybrid fold change)', fontsize = axis_fs)\n",
        "\n",
        "ax[1].scatter(P[random_idx]-H[random_idx],H[random_idx],s=s,alpha=alpha,  c = colors[random_idx])\n",
        "ax[1].set_xlabel('$R_{P} - R_{H}$',fontsize = axis_fs)\n",
        "ax[1].set_ylabel('$R_{H}$',fontsize = axis_fs)\n",
        "\n",
        "ax[2].scatter(cis_prop_reordered[random_idx], P[random_idx] ,s = s, alpha = alpha, c=colors[random_idx])\n",
        "# ax[2].set_xlabel(r'$\\frac{2}{\\pi} \\theta$ (scaled angle from $trans$ axis)',fontsize=axis_fs)\n",
        "ax[2].set_ylabel('$R_{P}$',fontsize=axis_fs)\n",
        "\n",
        "for i in range(2):\n",
        "    ax[i].set_xlabel(ax[i].get_xlabel(), fontsize=axis_fs)\n",
        "    ax[i].set_ylabel(ax[i].get_ylabel(), fontsize=axis_fs)\n",
        "    ax[i].tick_params(axis='both', labelsize=label_fs)\n",
        "    ax[i].grid(which='both', alpha=0.5, linewidth=0.5)\n",
        "    ax[i].minorticks_on()\n",
        "\n",
        "\n",
        "# Determine the diagonal points for the line y = x\n",
        "x_min, x_max = ax[i].get_xlim()\n",
        "y_min, y_max = ax[i].get_ylim()\n",
        "diag_val = max( np.abs(x_min), np.abs(x_max), np.abs(y_min), np.abs(y_max) )\n",
        "\n",
        "\n",
        "x = np.linspace(-diag_val, diag_val, 100)\n",
        "ax_ = ax[0]\n",
        "ax_.plot(x, x, color='orangered', linewidth=axis_width,alpha = line_alpha,zorder=0)\n",
        "ax_.axhline(0, color='darkblue', linewidth=axis_width, alpha=line_alpha,zorder=0)\n",
        "ax_.plot(x+x, x, color='skyblue', linewidth=axis_width,alpha = line_alpha,zorder=0)\n",
        "ax_.plot(-x-x, x, color='forestgreen', linewidth=axis_width,alpha = line_alpha,zorder=0)\n",
        "# ax_.axhline(0, color='black', linewidth=2, alpha=0.7, linestyle='-')\n",
        "ax_.axvline(0, color='black', linewidth=2, alpha=0.7, linestyle='-',zorder=0)\n",
        "ax_.set_xlim(-diag_val, diag_val)\n",
        "ax_.set_ylim(-diag_val, diag_val)\n",
        "\n",
        "ax_ = ax[1]\n",
        "ax_.axvline(0, color='orangered', linewidth=axis_width, alpha=line_alpha,zorder=0)\n",
        "ax_.axhline(0, color='darkblue', linewidth=axis_width, alpha = line_alpha,zorder=0)\n",
        "ax_.plot(x, x, color='skyblue', linewidth=axis_width,alpha = line_alpha,zorder=0)\n",
        "ax_.plot(-x, x, color='forestgreen', linewidth=axis_width,alpha = line_alpha,zorder=0)\n",
        "ax_.set_xlim(-diag_val, diag_val)\n",
        "ax_.set_ylim(-diag_val, diag_val)\n",
        "\n",
        "# Set additional grid lines for the third plot\n",
        "ax_ = ax[2]\n",
        "ax_.axvline(0.5, color='orangered', linewidth=axis_width, alpha=line_alpha,zorder=0)\n",
        "ax_.axvline(-0.5, color='darkblue', linewidth=axis_width, alpha=line_alpha,zorder=0)\n",
        "ax_.axvline(0.0, color='skyblue', linewidth=axis_width, alpha=line_alpha,zorder=0)\n",
        "ax_.axvline(-1, color='forestgreen', linewidth=axis_width, alpha=line_alpha,zorder=0)\n",
        "ax_.axvline(1, color='forestgreen', linewidth=axis_width, alpha=line_alpha,zorder=0)\n",
        "ax_.axhline(0, color='gray', linewidth=axis_width, alpha = line_alpha,zorder=0)\n",
        "ax_.axhline(0, color='black', linewidth=2, alpha=0.7, linestyle='-')\n",
        "\n",
        "ax[2].tick_params(axis='both', labelsize=label_fs)\n",
        "ax[2].tick_params(bottom=False)\n",
        "ax[2].grid(which='both', alpha=0.5, linewidth=0.5)\n",
        "ax[2].minorticks_on()\n",
        "\n",
        "ax2_xlims = ax[2].get_xlim()\n",
        "ax2_ylims = ax[2].get_ylim()\n",
        "\n",
        "plt.tight_layout()\n",
        "plt.subplots_adjust(hspace=0.2)\n",
        "plt.savefig('figs/fig2A',dpi=450,bbox_inches='tight')\n",
        "plt.show();\n"
      ],
      "metadata": {
        "id": "TzvBh7CQ0CnF"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "\n",
        "\n",
        "num_per_reg_group_theirs = []\n",
        "colors = []\n",
        "\n",
        "for reg in reg_groups:\n",
        "    fig, ax = plt.subplots(3, 1, figsize=(6, 23*(3/4)))\n",
        "    label = label_dict[reg]\n",
        "    color = color_dict[reg]\n",
        "    df_ = df[df['new_reg_group_theirs'] == reg]\n",
        "    P = df_['Parlog2FC'].values.flatten()\n",
        "    H = df_['Hyblog2FC'].values.flatten()\n",
        "    num_per_reg_group_theirs.append(len(H))\n",
        "    colors.append(color)\n",
        "    print(reg,len(P))\n",
        "    delta = P - H\n",
        "    theta_scaled = (2/np.pi) * np.arctan(H / delta)\n",
        "    R = np.sqrt(delta**2 + H**2)\n",
        "    cis_prop_reordered = theta_scaled-0.5\n",
        "    cis_prop_reordered[cis_prop_reordered<=-1.0] = (cis_prop_reordered[cis_prop_reordered<=-1.0]) + 2.0\n",
        "\n",
        "    ax[0].scatter(P,H,s = s, alpha = alpha, color = color, label = label)\n",
        "    ax[0].set_xlabel('$R_{P}$ (log2 parental fold change)', fontsize = axis_fs)\n",
        "    ax[0].set_ylabel('$R_{H}$ (log2 hybrid fold change)', fontsize = axis_fs)\n",
        "\n",
        "    ax[1].scatter(P-H,H,s=s,alpha=alpha,  color = color)\n",
        "    ax[1].set_xlabel('$R_{P} - R_{H}$',fontsize = axis_fs)\n",
        "    ax[1].set_ylabel('$R_{H}$',fontsize = axis_fs)\n",
        "\n",
        "    ax[2].scatter(cis_prop_reordered, P ,s = s, alpha = alpha,color=color)\n",
        "    # ax[2].set_xlabel(r'$\\frac{2}{\\pi} \\theta$ (scaled angle from $trans$ axis)',fontsize=axis_fs)\n",
        "    ax[2].set_ylabel('$R_{P}$',fontsize=axis_fs)\n",
        "\n",
        "    for i in range(2):\n",
        "        ax[i].set_xlabel(ax[i].get_xlabel(), fontsize=axis_fs)\n",
        "        ax[i].set_ylabel(ax[i].get_ylabel(), fontsize=axis_fs)\n",
        "        ax[i].tick_params(axis='both', labelsize=label_fs)\n",
        "        ax[i].grid(which='both', alpha=0.5, linewidth=0.5)\n",
        "        ax[i].minorticks_on()\n",
        "\n",
        "        ax[i].set_xlim(-diag_val, diag_val)\n",
        "        ax[i].set_ylim(-diag_val, diag_val)\n",
        "\n",
        "\n",
        "    x = np.linspace(-diag_val, diag_val, 100)\n",
        "    ax_ = ax[0]\n",
        "    ax_.plot(x, x, color='orangered', linewidth=axis_width,alpha = line_alpha,zorder=0)\n",
        "    ax_.axhline(0, color='darkblue', linewidth=axis_width, alpha=line_alpha,zorder=0)\n",
        "    ax_.plot(x+x, x, color='skyblue', linewidth=axis_width,alpha = line_alpha,zorder=0)\n",
        "    ax_.plot(-x-x, x, color='forestgreen', linewidth=axis_width,alpha = line_alpha,zorder=0)\n",
        "    # ax_.axhline(0, color='black', linewidth=2, alpha=0.7, linestyle='-')\n",
        "    ax_.axvline(0, color='black', linewidth=2, alpha=0.7, linestyle='-',zorder=0)\n",
        "\n",
        "    ax_ = ax[1]\n",
        "    ax_.axvline(0, color='orangered', linewidth=axis_width, alpha=line_alpha,zorder=0)\n",
        "    ax_.axhline(0, color='darkblue', linewidth=axis_width, alpha = line_alpha,zorder=0)\n",
        "    ax_.plot(x, x, color='skyblue', linewidth=axis_width,alpha = line_alpha,zorder=0)\n",
        "    ax_.plot(-x, x, color='forestgreen', linewidth=axis_width,alpha = line_alpha,zorder=0)\n",
        "\n",
        "\n",
        "\n",
        "    # Set additional grid lines for the third plot\n",
        "    ax_ = ax[2]\n",
        "    ax_.axvline(0.5, color='orangered', linewidth=axis_width, alpha=line_alpha,zorder=0)\n",
        "    ax_.axvline(-0.5, color='darkblue', linewidth=axis_width, alpha=line_alpha,zorder=0)\n",
        "    ax_.axvline(0.0, color='skyblue', linewidth=axis_width, alpha=line_alpha,zorder=0)\n",
        "    ax_.axvline(-1, color='forestgreen', linewidth=axis_width, alpha=line_alpha,zorder=0)\n",
        "    ax_.axvline(1, color='forestgreen', linewidth=axis_width, alpha=line_alpha,zorder=0)\n",
        "    ax_.axhline(0, color='gray', linewidth=axis_width, alpha = line_alpha,zorder=0)\n",
        "    ax_.axhline(0, color='black', linewidth=2, alpha=0.7, linestyle='-')\n",
        "\n",
        "    ax[2].tick_params(axis='both', labelsize=label_fs)\n",
        "    ax[2].tick_params(bottom=False)\n",
        "    ax[2].grid(which='both', alpha=0.5, linewidth=0.5)\n",
        "    ax[2].minorticks_on()\n",
        "\n",
        "    ax[2].set_xlim(ax2_xlims)\n",
        "    ax[2].set_ylim(ax2_ylims)\n",
        "\n",
        "    plt.tight_layout()\n",
        "    plt.subplots_adjust(hspace=0.2)\n",
        "    plt.savefig(f'supp2B_{reg}',dpi=450,bbox_inches='tight')\n",
        "    plt.show();"
      ],
      "metadata": {
        "id": "1T7ES1SA0FSC"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "\n",
        "## Transformed regulatory assignments\n",
        "\n",
        "# random order\n",
        "fig, ax = plt.subplots(3, 1, figsize=(6, 23*(3/4)))\n",
        "\n",
        "s = 2\n",
        "alpha = 1.0\n",
        "axis_fs = 25\n",
        "title_fs = 18\n",
        "label_fs = 20\n",
        "line_color = 'black'\n",
        "axis_width = 3.0\n",
        "\n",
        "\n",
        "P = df['Parlog2FC'].values.flatten()\n",
        "H = df['Hyblog2FC'].values.flatten()\n",
        "delta = P - H\n",
        "theta_scaled = (2/np.pi) * np.arctan(H / delta)\n",
        "R = np.sqrt(delta**2 + H**2)\n",
        "cis_prop_reordered = theta_scaled-0.5\n",
        "cis_prop_reordered[cis_prop_reordered<=-1.0] = (cis_prop_reordered[cis_prop_reordered<=-1.0]) + 2.0\n",
        "\n",
        "colors = df['color_fdr'].values\n",
        "\n",
        "random_idx = np.arange(0,len(colors))\n",
        "np.random.shuffle(random_idx)\n",
        "\n",
        "\n",
        "ax[0].scatter(P[random_idx],H[random_idx],s = s, alpha = alpha, c = colors[random_idx])\n",
        "ax[0].set_xlabel('$R_{P}$ (log2 parental fold change)', fontsize = axis_fs)\n",
        "ax[0].set_ylabel('$R_{H}$ (log2 hybrid fold change)', fontsize = axis_fs)\n",
        "\n",
        "ax[1].scatter(P[random_idx]-H[random_idx],H[random_idx],s=s,alpha=alpha,  c = colors[random_idx])\n",
        "ax[1].set_xlabel('$R_{P} - R_{H}$',fontsize = axis_fs)\n",
        "ax[1].set_ylabel('$R_{H}$',fontsize = axis_fs)\n",
        "\n",
        "ax[2].scatter(cis_prop_reordered[random_idx], P[random_idx] ,s = s, alpha = alpha, c=colors[random_idx])\n",
        "# ax[2].set_xlabel(r'$\\frac{2}{\\pi} \\theta$ (scaled angle from $trans$ axis)',fontsize=axis_fs)\n",
        "ax[2].set_ylabel('$R_{P}$',fontsize=axis_fs)\n",
        "\n",
        "for i in range(2):\n",
        "    ax[i].set_xlabel(ax[i].get_xlabel(), fontsize=axis_fs)\n",
        "    ax[i].set_ylabel(ax[i].get_ylabel(), fontsize=axis_fs)\n",
        "    ax[i].tick_params(axis='both', labelsize=label_fs)\n",
        "    ax[i].grid(which='both', alpha=0.5, linewidth=0.5)\n",
        "    ax[i].minorticks_on()\n",
        "\n",
        "\n",
        "# Determine the diagonal points for the line y = x\n",
        "x_min, x_max = ax[i].get_xlim()\n",
        "y_min, y_max = ax[i].get_ylim()\n",
        "diag_val = max( np.abs(x_min), np.abs(x_max), np.abs(y_min), np.abs(y_max) )\n",
        "\n",
        "\n",
        "\n",
        "x = np.linspace(-diag_val, diag_val, 100)\n",
        "ax_ = ax[0]\n",
        "ax_.plot(x, x, color='orangered', linewidth=axis_width,alpha = line_alpha,zorder=0)\n",
        "ax_.axhline(0, color='darkblue', linewidth=axis_width, alpha=line_alpha,zorder=0)\n",
        "ax_.plot(x+x, x, color='skyblue', linewidth=axis_width,alpha = line_alpha,zorder=0)\n",
        "ax_.plot(-x-x, x, color='forestgreen', linewidth=axis_width,alpha = line_alpha,zorder=0)\n",
        "# ax_.axhline(0, color='black', linewidth=2, alpha=0.7, linestyle='-')\n",
        "ax_.axvline(0, color='black', linewidth=2, alpha=0.7, linestyle='-',zorder=0)\n",
        "ax_.set_xlim(-diag_val, diag_val)\n",
        "ax_.set_ylim(-diag_val, diag_val)\n",
        "\n",
        "ax_ = ax[1]\n",
        "ax_.axvline(0, color='orangered', linewidth=axis_width, alpha=line_alpha,zorder=0)\n",
        "ax_.axhline(0, color='darkblue', linewidth=axis_width, alpha = line_alpha,zorder=0)\n",
        "ax_.plot(x, x, color='skyblue', linewidth=axis_width,alpha = line_alpha,zorder=0)\n",
        "ax_.plot(-x, x, color='forestgreen', linewidth=axis_width,alpha = line_alpha,zorder=0)\n",
        "ax_.set_xlim(-diag_val, diag_val)\n",
        "ax_.set_ylim(-diag_val, diag_val)\n",
        "\n",
        "# Set additional grid lines for the third plot\n",
        "ax_ = ax[2]\n",
        "ax_.axvline(0.5, color='orangered', linewidth=axis_width, alpha=line_alpha,zorder=0)\n",
        "ax_.axvline(-0.5, color='darkblue', linewidth=axis_width, alpha=line_alpha,zorder=0)\n",
        "ax_.axvline(0.0, color='skyblue', linewidth=axis_width, alpha=line_alpha,zorder=0)\n",
        "ax_.axvline(-1, color='forestgreen', linewidth=axis_width, alpha=line_alpha,zorder=0)\n",
        "ax_.axvline(1, color='forestgreen', linewidth=axis_width, alpha=line_alpha,zorder=0)\n",
        "ax_.axhline(0, color='gray', linewidth=axis_width, alpha = line_alpha,zorder=0)\n",
        "ax_.axhline(0, color='black', linewidth=2, alpha=0.7, linestyle='-')\n",
        "\n",
        "ax[2].tick_params(axis='both', labelsize=label_fs)\n",
        "ax[2].tick_params(bottom=False)\n",
        "ax[2].grid(which='both', alpha=0.5, linewidth=0.5)\n",
        "ax[2].minorticks_on()\n",
        "\n",
        "\n",
        "ax2_xlims = ax[2].get_xlim()\n",
        "ax2_ylims = ax[2].get_ylim()\n",
        "\n",
        "plt.tight_layout()\n",
        "plt.subplots_adjust(hspace=0.2)\n",
        "plt.savefig('figs/fig2C',dpi=450,bbox_inches='tight')\n",
        "plt.show();\n",
        "\n"
      ],
      "metadata": {
        "id": "Cull9SfT0JLj"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "\n",
        "s = 2\n",
        "alpha = 1.0\n",
        "axis_fs = 25\n",
        "title_fs = 18\n",
        "label_fs = 20\n",
        "line_color = 'black'\n",
        "axis_width = 3.0\n",
        "line_alpha = 0.7\n",
        "\n",
        "num_per_reg_group_theirs = []\n",
        "colors = []\n",
        "\n",
        "for reg in reg_groups:\n",
        "    fig, ax = plt.subplots(3, 1, figsize=(6, 23*(3/4)))\n",
        "    label = label_dict[reg]\n",
        "    color = color_dict[reg]\n",
        "    df_ = df[df['new_reg_group_fdr'] == reg]\n",
        "    P = df_['Parlog2FC'].values.flatten()\n",
        "    H = df_['Hyblog2FC'].values.flatten()\n",
        "    num_per_reg_group_theirs.append(len(H))\n",
        "    colors.append(color)\n",
        "    print(reg,len(P))\n",
        "    delta = P - H\n",
        "    theta_scaled = (2/np.pi) * np.arctan(H / delta)\n",
        "    R = np.sqrt(delta**2 + H**2)\n",
        "    cis_prop_reordered = theta_scaled-0.5\n",
        "    cis_prop_reordered[cis_prop_reordered<=-1.0] = (cis_prop_reordered[cis_prop_reordered<=-1.0]) + 2.0\n",
        "\n",
        "    # ax[0].scatter(P, H, s=s, alpha=alpha, color=color, label=label)\n",
        "    # ax[1].scatter(delta, H, s=s, alpha=alpha, color=color)\n",
        "    # ax[2].scatter(theta_scaled, P, s=s, alpha=alpha, color=color)\n",
        "    # ax[2].scatter(theta_scaled, R, s=s, alpha=alpha, color=color)\n",
        "\n",
        "    ax[0].scatter(P,H,s = s, alpha = alpha, color = color, label = label)\n",
        "    ax[0].set_xlabel('$R_{P}$ (log2 parental fold change)', fontsize = axis_fs)\n",
        "    ax[0].set_ylabel('$R_{H}$ (log2 hybrid fold change)', fontsize = axis_fs)\n",
        "\n",
        "    ax[1].scatter(P-H,H,s=s,alpha=alpha,  color = color)\n",
        "    ax[1].set_xlabel('$R_{P} - R_{H}$',fontsize = axis_fs)\n",
        "    ax[1].set_ylabel('$R_{H}$',fontsize = axis_fs)\n",
        "\n",
        "    ax[2].scatter(cis_prop_reordered, P ,s = s, alpha = alpha,color=color)\n",
        "    # ax[2].set_xlabel(r'$\\frac{2}{\\pi} \\theta$ (scaled angle from $trans$ axis)',fontsize=axis_fs)\n",
        "    ax[2].set_ylabel('$R_{P}$',fontsize=axis_fs)\n",
        "\n",
        "    for i in range(2):\n",
        "        ax[i].set_xlabel(ax[i].get_xlabel(), fontsize=axis_fs)\n",
        "        ax[i].set_ylabel(ax[i].get_ylabel(), fontsize=axis_fs)\n",
        "        ax[i].tick_params(axis='both', labelsize=label_fs)\n",
        "        ax[i].grid(which='both', alpha=0.5, linewidth=0.5)\n",
        "        ax[i].minorticks_on()\n",
        "\n",
        "\n",
        "        # Determine the diagonal points for the line y = x\n",
        "\n",
        "        # ax[i].plot([-diag_val, diag_val], np.sqrt(3)*np.array([-diag_val, diag_val]), color='black',alpha=0.7,linestyle='--')\n",
        "        # ax[i].plot([-diag_val, diag_val], np.sqrt(3)*np.array([diag_val, -diag_val]), color='black',alpha=0.7,linestyle='--')\n",
        "        # ax[i].plot([-diag_val, diag_val], (1/np.sqrt(3))*np.array([-diag_val, diag_val]), color='black',alpha=0.7,linestyle='--')\n",
        "        # ax[i].plot([-diag_val, diag_val], (1/np.sqrt(3))*np.array([diag_val, -diag_val]), color='black',alpha=0.7,linestyle='--')\n",
        "        ax[i].set_xlim(-diag_val, diag_val)\n",
        "        ax[i].set_ylim(-diag_val, diag_val)\n",
        "\n",
        "\n",
        "    x = np.linspace(-diag_val, diag_val, 100)\n",
        "    ax_ = ax[0]\n",
        "    ax_.plot(x, x, color='orangered', linewidth=axis_width,alpha = line_alpha,zorder=0)\n",
        "    ax_.axhline(0, color='darkblue', linewidth=axis_width, alpha=line_alpha,zorder=0)\n",
        "    ax_.plot(x+x, x, color='skyblue', linewidth=axis_width,alpha = line_alpha,zorder=0)\n",
        "    ax_.plot(-x-x, x, color='forestgreen', linewidth=axis_width,alpha = line_alpha,zorder=0)\n",
        "    # ax_.axhline(0, color='black', linewidth=2, alpha=0.7, linestyle='-')\n",
        "    ax_.axvline(0, color='black', linewidth=2, alpha=0.7, linestyle='-',zorder=0)\n",
        "\n",
        "    ax_ = ax[1]\n",
        "    ax_.axvline(0, color='orangered', linewidth=axis_width, alpha=line_alpha,zorder=0)\n",
        "    ax_.axhline(0, color='darkblue', linewidth=axis_width, alpha = line_alpha,zorder=0)\n",
        "    ax_.plot(x, x, color='skyblue', linewidth=axis_width,alpha = line_alpha,zorder=0)\n",
        "    ax_.plot(-x, x, color='forestgreen', linewidth=axis_width,alpha = line_alpha,zorder=0)\n",
        "\n",
        "\n",
        "\n",
        "    # Set additional grid lines for the third plot\n",
        "    ax_ = ax[2]\n",
        "    ax_.axvline(0.5, color='orangered', linewidth=axis_width, alpha=line_alpha,zorder=0)\n",
        "    ax_.axvline(-0.5, color='darkblue', linewidth=axis_width, alpha=line_alpha,zorder=0)\n",
        "    ax_.axvline(0.0, color='skyblue', linewidth=axis_width, alpha=line_alpha,zorder=0)\n",
        "    ax_.axvline(-1, color='forestgreen', linewidth=axis_width, alpha=line_alpha,zorder=0)\n",
        "    ax_.axvline(1, color='forestgreen', linewidth=axis_width, alpha=line_alpha,zorder=0)\n",
        "    ax_.axhline(0, color='gray', linewidth=axis_width, alpha = line_alpha,zorder=0)\n",
        "    ax_.axhline(0, color='black', linewidth=2, alpha=0.7, linestyle='-')\n",
        "\n",
        "    ax[2].tick_params(axis='both', labelsize=label_fs)\n",
        "    ax[2].tick_params(bottom=False)\n",
        "    ax[2].grid(which='both', alpha=0.5, linewidth=0.5)\n",
        "    ax[2].minorticks_on()\n",
        "\n",
        "    ax[2].set_xlim(ax2_xlims)\n",
        "    ax[2].set_ylim(ax2_ylims)\n",
        "\n",
        "    plt.tight_layout()\n",
        "    plt.subplots_adjust(hspace=0.2)\n",
        "    plt.savefig(f'supp2B_{reg}_fdr',dpi=450,bbox_inches='tight')\n",
        "    plt.show();\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "V9akxcfezbk_"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "\n",
        "## Genes per regulatory assignment\n",
        "\n",
        "num_per_reg_group_theirs = []\n",
        "num_per_reg_group_ours = []\n",
        "\n",
        "for reg in reg_groups:\n",
        "    num_per_reg_group_theirs.append(int(sum(df['new_reg_group_theirs']==reg)))\n",
        "    num_per_reg_group_ours.append(int(sum(df['new_reg_group_fdr']==reg)))\n",
        "\n",
        "colors = [color_dict[reg] for reg in reg_groups]\n",
        "\n",
        "# barplot\n",
        "\n",
        "fig, ax = plt.subplots(1,3,figsize = (5.75*3,6) )\n",
        "\n",
        "axis_fs = 20\n",
        "title_fs = 20\n",
        "i = 0\n",
        "bars = ax[i].bar(range(len(num_per_reg_group_theirs)), num_per_reg_group_theirs, color=colors)\n",
        "\n",
        "for bar in bars:\n",
        "  height = bar.get_height()\n",
        "  ax[i].text(bar.get_x() + bar.get_width() / 2, height, f'{height:.0f}',\n",
        "             ha='center', va='bottom', fontsize=14)\n",
        "\n",
        "ax[i].set_xticks(range(len(num_per_reg_group_theirs)))\n",
        "ax[i].set_xticklabels(reg_groups,fontsize=axis_fs-4)\n",
        "ax[i].set_xlabel('Assigned regulation group',fontsize=axis_fs)\n",
        "ax[i].set_ylabel('Number of genes',fontsize= axis_fs)\n",
        "ax[i].set_yscale('log')\n",
        "ax[i].yaxis.grid(which='both', alpha=0.5, linewidth=0.5)\n",
        "ax[i].minorticks_on()\n",
        "ax[i].xaxis.set_tick_params(which='minor', bottom=False)\n",
        "ax[i].set_title('Untransformed assignments',fontsize=title_fs)\n",
        "i = 1\n",
        "bars = ax[i].bar(range(len(num_per_reg_group_ours)), num_per_reg_group_ours, color=colors)\n",
        "\n",
        "for bar in bars:\n",
        "  height = bar.get_height()\n",
        "  ax[i].text(bar.get_x() + bar.get_width() / 2, height, f'{height:.0f}',\n",
        "             ha='center', va='bottom',fontsize=14)\n",
        "\n",
        "ax[i].set_xticks(range(len(num_per_reg_group_ours)))\n",
        "ax[i].set_xticklabels(reg_groups,fontsize=axis_fs-4)\n",
        "ax[i].set_xlabel('Assigned regulation group',fontsize=axis_fs)\n",
        "ax[i].set_ylabel('Number of genes',fontsize= axis_fs)\n",
        "ax[i].set_yscale('log')\n",
        "ax[i].yaxis.grid(which='both', alpha=0.5, linewidth=0.5)\n",
        "ax[i].minorticks_on()\n",
        "ax[i].xaxis.set_tick_params(which='minor', bottom=False)\n",
        "ax[i].set_title('Transformed assignments',fontsize=title_fs)\n",
        "ax[i].set_ylim(ax[0].get_ylim())\n",
        "\n",
        "i = 2\n",
        "for r,reg in enumerate(reg_groups):\n",
        "\n",
        "    ax[i].scatter(num_per_reg_group_theirs[r],num_per_reg_group_ours[r],\n",
        "                  color = colors[r],s=160,alpha=0.8,\n",
        "                  edgecolor='black',label=reg)\n",
        "\n",
        "ax[i].set_xscale('log')\n",
        "ax[i].set_yscale('log')\n",
        "ax[i].set_xlabel('Untransformed assignments',fontsize=axis_fs)\n",
        "ax[i].set_ylabel('Transformed assignments',fontsize=axis_fs)\n",
        "ax[i].set_title('Assignment strategies compared',fontsize=axis_fs)\n",
        "ax[i].yaxis.grid(which='both', alpha=0.5, linewidth=0.5)\n",
        "ax[i].minorticks_on()\n",
        "lims = ax[i].get_xlim()\n",
        "ax[i].plot(lims,lims,color='black',linestyle='--',label='Identity')\n",
        "ax[i].legend(fontsize=15)\n",
        "\n",
        "\n",
        "ax[i].set_ylim(lims)\n",
        "ax[i].set_xlim(lims)\n",
        "\n",
        "\n",
        "for ax_ in ax:\n",
        "    ax_.tick_params(axis='x', which='major', labelsize=11)\n",
        "    ax_.tick_params(axis='y', which='major', labelsize=20)\n",
        "ax[2].tick_params(axis='both', which='major', labelsize=20)\n",
        "\n",
        "plt.tight_layout()\n",
        "plt.savefig('./figs/supp_reg_assignments_bar',dpi=450,bbox_inches='tight')\n",
        "plt.show()"
      ],
      "metadata": {
        "id": "j7_eJZFm0P_7"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "------\n",
        "\n",
        "# Alluvian plot"
      ],
      "metadata": {
        "id": "KMtte5lwzl4F"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "# download from https://github.com/vinsburg/alluvial_diagram/tree/master\n",
        "!wget https://raw.githubusercontent.com/vinsburg/alluvial_diagram/master/alluvial.py\n"
      ],
      "metadata": {
        "id": "BBaeqZFbzp7J"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import alluvial"
      ],
      "metadata": {
        "id": "vKim2tZczrRt"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "num_overlap_dict = {r : {} for r in reg_groups}\n",
        "\n",
        "log_num_overlap_dict = {r : {} for r in reg_groups}\n",
        "\n",
        "for r1 in reg_groups:\n",
        "  their_indices = df[df['new_reg_group_theirs'] == r1].index.values.flatten()\n",
        "  for r2 in reg_groups:\n",
        "    our_indices = df[df['new_reg_group_fdr'] == r2 ].index.values.flatten()\n",
        "    num_overlap = len(set(their_indices).intersection(set(our_indices)))\n",
        "    num_overlap_dict[r1][r2+' * '] = num_overlap\n",
        "    if num_overlap!= 0:\n",
        "      log_num_overlap_dict[r1][r2+' * '] = np.log(num_overlap)\n",
        "    else:\n",
        "      log_num_overlap_dict[r1][r2+' * '] = 0\n",
        "\n",
        "\n",
        "alluvial_colors = ['skyblue','orangered','forestgreen','royalblue','lightgray']\n",
        "ax = alluvial.plot(num_overlap_dict,colors=alluvial_colors,alpha = 0.9)\n",
        "fig = ax.get_figure()\n",
        "fig.set_size_inches(2,8)\n",
        "plt.savefig('alluvian',dpi=450,bbox_inches='tight')\n",
        "plt.show()\n"
      ],
      "metadata": {
        "id": "6mb6Vq3ezjfz"
      },
      "execution_count": null,
      "outputs": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyMhwyNUHUxAEPq9mYAa0Byz",
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}